File: func11.js
// func11

const Box = x => ({
    map: f => Box(f(x)),
    fold: f => f(x),
    inspect: () => `Box(${x})`
})

const LazyBox = g => ({
    map: f => LazyBox(() => f(g())),
    fold: f => f(g())
})

const nextCharForNumberString = str =>
    LazyBox(() => str)
    .map(s => s.trim())
    .map(r => parseInt(r))
    .map(i => i + 1)
    .fold(j => String.fromCharCode(j)) // if fold is not called, nothing runs



const result = nextCharForNumberString(' 64 ');

console.log(result)

File: func11.py
# func11

class Box:
    def __init__(self, x):
        self.x = x
        self.map = lambda f: Box(f(x))
        self.fold = lambda f: f(x)
    
    def __repr__(self):
        return 'Box({})'.format(self.x)

# Box = lambda x: {
#     'map': lambda f: Box(f(x))
#     'fold': lambda f: f(x)
# }

# def nextCharForNumberString(st):
#     return map(lambda i: chr(i), map(lambda n: n+1, map(lambda x: int(x), map(lambda s: s.strip(), [st]))))

class LazyBox:
    def __init__(self, g):
        self.map = lambda f: LazyBox(lambda: f(g()))
        self.fold = lambda f: f(g())

def intify(x):
    print("intifying")
    return int(x)

def nextCharForNumberString(st):
    return LazyBox(lambda: st).map(lambda s: s.strip()) \
        .map(intify) \
        .map(lambda n: n+1) \
        .fold(lambda i: chr(i)) # intify will not be called if fold is commented out


result = nextCharForNumberString(' 64 ')

print(result)